README.md
Postfix Dovecot Cookbook
========================
[![GitHub](http://img.shields.io/badge/github-zuazo/postfix--dovecot--cookbook-blue.svg?style=flat)](https://github.com/zuazo/postfix-dovecot-cookbook)
[![License](https://img.shields.io/github/license/zuazo/postfix-dovecot-cookbook.svg?style=flat)](#license-and-author)

[![Cookbook Version](https://img.shields.io/cookbook/v/postfix-dovecot.svg?style=flat)](https://supermarket.chef.io/cookbooks/postfix-dovecot)
[![Dependency Status](http://img.shields.io/gemnasium/zuazo/postfix-dovecot-cookbook.svg?style=flat)](https://gemnasium.com/zuazo/postfix-dovecot-cookbook)
[![Build Status](http://img.shields.io/travis/zuazo/postfix-dovecot-cookbook/3.0.0.svg?style=flat)](https://travis-ci.org/zuazo/postfix-dovecot-cookbook)

Installs and configures a mail server using [Postfix](http://www.postfix.org/), [Dovecot](http://www.dovecot.org/), [PostfixAdmin](http://postfixadmin.sourceforge.net/) and [SpamAssassin](http://spamassassin.apache.org/), including [Amazon SES](http://aws.amazon.com/ses/) support.

Requirements
============

## Supported Platforms

This cookbook has been tested on the following platforms:

* Amazon Linux
* CentOS `>= 6.0`
* Debian `>= 7.0`
* Fedora `>= 17.0`
* Ubuntu `>= 12.04`

Please, [let us know](https://github.com/zuazo/postfix-dovecot-cookbook/issues/new?title=I%20have%20used%20it%20successfully%20on%20...) if you use it successfully on any other platform.

## Required Cookbooks

* [chef-vault](https://supermarket.chef.io/cookbooks/chef-vault)
* [build-essential](https://supermarket.chef.io/cookbooks/build-essential)
* [dovecot](https://supermarket.chef.io/cookbooks/dovecot)
* [onddo-spamassassin](https://supermarket.chef.io/cookbooks/onddo-spamassassin)
* [postfixadmin](https://supermarket.chef.io/cookbooks/postfixadmin)
* [postfix-full](https://supermarket.chef.io/cookbooks/postfix-full) by [Malte Swart](https://github.com/mswart)
* [ssl_certificate](https://supermarket.chef.io/cookbooks/ssl_certificate)
* [yum](https://supermarket.chef.io/cookbooks/yum)

## Required Applications

* Dovecot `>= 2`: requires this version of dovecot to be available by the distribution's package manager
* Chef `12.5` or higher.
* Ruby `2.2` or higher.

Attributes
==========

| Attribute                                         | Default                | Description                       |
|:--------------------------------------------------|:-----------------------|:----------------------------------|
| `node['postfix-dovecot']['postmaster_address']`   | `'postmaster@foo.bar'` | Postmaster mail address.
| `node['postfix-dovecot']['hostname']`             | `node['fqdn']`         | Hostname.
| `node['postfix-dovecot']['rbls']`                 | `[]`                   | Mail RBLs array.
| `node['postfix-dovecot']['database']['type']`     | `'mysql'`              | Database type. Possible values are: `'mysql'`, `'postgresql'` (Please, see [below](#postgresql-support)).
| `node['postfix-dovecot']['sieve']['enabled']`     | `true`                 | Whether to enable sieve.
| `node['postfix-dovecot']['sieve']['global_path']` | `"#{default['dovecot']['conf_path']}/sieve/default.sieve"` | Sieve global path.
| `node['postfix-dovecot']['spamc']['enabled']`     | `true`                 | Whether to enable SpamAssassin.
| `node['postfix-dovecot']['spamc']['recipe']`      | `'onddo-spamassassin'` | Spamc recipe name to use.
| `node['postfix-dovecot']['vmail']['user']`        | `'vmail'`              | Virtual mail system user name.
| `node['postfix-dovecot']['vmail']['group']`       | `node['postfix-dovecot']['vmail']['user']` | Virtual mail system group name.
| `node['postfix-dovecot']['vmail']['uid']`         | `5000`                 | Virtual mail system user id.
| `node['postfix-dovecot']['vmail']['gid']`         | `node['postfix-dovecot']['vmail']['uid']` | Virtual mail system group id.
| `node['postfix-dovecot']['vmail']['home']`        | `'/var/vmail'`         | Virtual mail user home path.

## Amazon SES Attributes

You can use `node['postfix-dovecot']['ses']['enabled']` to enable SES for sending emails.

| Attribute                                    | Default        | Description                       |
|:---------------------------------------------|:---------------|:----------------------------------|
| `node['postfix-dovecot']['ses']['enabled']`  | `false`        | Whether to enable [Amazon SES](http://aws.amazon.com/ses/).
| `node['postfix-dovecot']['ses']['source']`   | `'attributes'` | Where to read the credentials from. Possible values: `'attributes'`,  `'chef-vault'`.
| `node['postfix-dovecot']['ses']['vault']`    | `'amazon'`     | Chef Vault bag to read SES credentials from.
| `node['postfix-dovecot']['ses']['item']`     | `'ses'`        | Chef Vault item.
| `node['postfix-dovecot']['ses']['region']`   | `'us-east-1'`  | Amazon AWS region, used to calculate the servers.
| `node['postfix-dovecot']['ses']['servers']`  | *calculated*   | Amazon SES SMTP servers array.
| `node['postfix-dovecot']['ses']['username']` | `'USERNAME'`   | SES SMTP username. See [Obtaining Your Amazon SES SMTP Credentials](http://docs.aws.amazon.com/ses/latest/DeveloperGuide/smtp-credentials.html) documentation.
| `node['postfix-dovecot']['ses']['password']` | `'PASSWORD'`   | Amazon SES SMTP password.

When Chef Vault is disabled in `node['postfix-dovecot']['ses']['source']`, this is the default behavior, the credentials are read from `['username']` and `['password']` attributes.

When credentials should be read using `chef-vault`, the Chef Vault bag must have the following structure:

```json
{
  "username": "AMAZON_USERNAME",
  "password": "AMAZON_PASSWORD"
}
```

See the [Chef-Vault documentation](https://github.com/Nordstrom/chef-vault/blob/master/README.md) to learn how to create chef-vault bags.

## The SSL Certificate

This cookbook uses the [`ssl_certificate`](https://supermarket.chef.io/cookbooks/ssl_certificate) cookbook to create the SSL certificate. The namespace used is `node['postfix-dovecot']`. For example:

```ruby
node.default['postfix-dovecot']['common_name'] = 'mail.example.com'
include_recipe 'postfix-dovecot'
```

This certificate is used for Postfix and Dovecot. For PostfixAdmin, you should use the `node['postfixadmin']` namespace.

You can also tweak the supported SSL ciphers [setting the `node['ssl_certificate']['service']['compatibility']` attribute](https://github.com/zuazo/ssl_certificate-cookbook#securing-server-side-tls):

```ruby
node.default['ssl_certificate']['service']['compatibility'] = :modern

include_recipe 'postfix-dovecot'
```

See the [`ssl_certificate` namespace documentation](https://supermarket.chef.io/cookbooks/ssl_certificate#namespaces) for more information.

Recipes
=======

## postfix-dovecot::default

Installs and configures everything.

## postfix-dovecot::vmail

Creates vmail user.

## postfix-dovecot::spam

Installs and configures SpamAssassin.

## postfix-dovecot::postfix

Installs and configures Postfix.

## postfix-dovecot::postfix_mysql

Installs Postfix package with MySQL support. Used by the `postfix-dovecot::postfix` recipe.

## postfix-dovecot::postfix_postgresql

Installs Postfix package with PostgreSQL support. Used by the `postfix-dovecot::postfix` recipe.

## postfix-dovecot::postfixadmin

Installs and configures PostfixAdmin.

## postfix-dovecot::dovecot

Installs and configures Dovecot 2.

Usage Examples
==============

## Including in a Cookbook Recipe

Running it from a recipe:

```ruby
node['postfix-dovecot']['postmaster_address'] = 'postmaster@foobar.com'
node['postfix-dovecot']['hostname'] = 'mail.foobar.com'

include_recipe 'postfix-dovecot::default'

postfixadmin_admin 'admin@admindomain.com' do
  password 'sup3r-s3cr3t-p4ss'
  action :create
end

postfixadmin_domain 'foobar.com' do
  login_username 'admin@admindomain.com'
  login_password 'sup3r-s3cr3t-p4ss'
end

postfixadmin_mailbox 'bob@foobar.com' do
  password 'alice'
  login_username 'admin@admindomain.com'
  login_password 'sup3r-s3cr3t-p4ss'
end

postfixadmin_alias 'billing@foobar.com' do
  goto 'bob@foobar.com'
  login_username 'admin@admindomain.com'
  login_password 'sup3r-s3cr3t-p4ss'
end
```

Don't forget to include the `postfix-dovecot` cookbook as a dependency in the metadata.

```ruby
# metadata.rb
# [...]

depends 'postfix-dovecot'
```

## Including in the Run List

Another alternative is to include the default recipe in your *Run List*.

```json
{
  "name": "mail.example.com",
  "[...]": "[...]",
  "run_list": [
    "[...]",
    "recipe[postfix-dovecot]"
  ]
}
```

## Enabling Some RBLs

You can enable some [RBLs](http://en.wikipedia.org/wiki/DNSBL) to avoid spam:

```ruby
node.default['postfix-dovecot']['rbls'] = %w(
  dnsbl.sorbs.net
  zen.spamhaus.org
  bl.spamcop.net
  cbl.abuseat.org
)
include_recipe 'postfix-dovecot::default'
```

PostgreSQL Support
==================

PostgreSQL support should be considered **experimental** at the moment. Use at your own risk.

[Any feedback you can provide regarding the PostgreSQL support](https://github.com/zuazo/postfix-dovecot-cookbook/issues/new?title=PostgreSQL%20Support) will be greatly appreciated.

## PostgreSQL Support on CentOS and Fedora

The latest CentOS and Fedora versions come without PostgreSQL support in their Postfix package. So we need to recompile it using the SRPM, enabling the PostgreSQL support.

The `postfix-dovecot::postfix_postgresql` recipe takes care of it transparently. This recipe has been tested using `test-kitchen`, but it may not work for all cases. This code has been tested in the following platforms:

* CentOS `6.5` and `7.0`
* Fedora `19` and `20`.

Please, [let us know](https://github.com/zuazo/postfix-dovecot-cookbook/issues/new?title=I%20have%20tested%20PostgreSQL%20support%20successfully%20on%20...) if you use PostgreSQL support successfully on any other platform.

## PostgreSQL Support on Amazon Linux

Support for PostgreSQL on Amazon Linux is still not finished because of the need to patch the provided SRPM. Its implementation would require a little monkey-patching.

Please, open an issue if [you need the support of PostgreSQL on Amazon Linux](https://github.com/zuazo/postfix-dovecot-cookbook/issues/new?title=We%20need%20PostgreSQL%20support%20on%20Amazon%20Linux).

## PostgreSQL Versions < 9.3

If you are using PostgreSQL version `< 9.3`, you may need to adjust the `shmmax` and `shmall` kernel parameters to configure the shared memory. You can see [the example used for the integration tests](test/cookbooks/postfix-dovecot_test/recipes/postgresql_memory.rb).

## PostgreSQL Support Related Attributes

Some cookbook attributes are used internally to add PostgreSQL support. They can make your journey smoother if you need to improve PostgreSQL support.

| Attribute                                                      | Default      | Description                    |
|:---------------------------------------------------------------|:-------------|:-------------------------------|
| `node['postfix-dovecot']['yum']`                               | *calculated* | A list of yum repositories to add to include the source SRPMs.
| `node['postfix-dovecot']['postfix']['srpm']['packages']`       | *calculated* | Packages required for compiling Postfix from sources.
| `node['postfix-dovecot']['postfix']['srpm']['rpm_regexp']`     | *calculated* | An array with two values, a pattern and a replacement. This Regexp is used to get the final Postfix RPM name from the SRPM name.
| `node['postfix-dovecot']['postfix']['srpm']['rpm_build_args']` | *calculated* | A string with the arguments to pass to *rpmbuild* application. Normally contains the required option to enable PostgreSQL in the Postfix SRPM.

See the [attributes/postfix_postgresql.rb](/attributes/postfix_postgresql.rb) file for default examples.

Please do not hesitate [to make a PR](https://github.com/zuazo/postfix-dovecot-cookbook/blob/master/TESTING.md) if you improve the PostgreSQL support ;-)

Testing
=======

See [TESTING.md](https://github.com/zuazo/postfix-dovecot-cookbook/blob/master/TESTING.md).

Contributing
============

Please do not hesitate to [open an issue](https://github.com/zuazo/postfix-dovecot-cookbook/issues/new) with any questions or problems.

See [CONTRIBUTING.md](https://github.com/zuazo/postfix-dovecot-cookbook/blob/master/CONTRIBUTING.md).

TODO
====

See [TODO.md](https://github.com/zuazo/postfix-dovecot-cookbook/blob/master/TODO.md).


License and Author
==================

|                      |                                          |
|:---------------------|:-----------------------------------------|
| **Author:**          | [Xabier de Zuazo](https://github.com/zuazo) (<xabier@zuazo.org>)
| **Contributor:**     | [Uwe Stuehler](https://github.com/ustuehler)
| **Copyright:**       | Copyright (c) 2015, Xabier de Zuazo
| **Copyright:**       | Copyright (c) 2014-2015, Onddo Labs, SL.
| **License:**         | Apache License, Version 2.0

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at
    
        http://www.apache.org/licenses/LICENSE-2.0
    
    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.


CHANGELOG.md
# Change Log
All notable changes to the `postfix-dovecot` cookbook will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
- ...

## [3.0.0] - 2017-03-14
### Added
- metadata: Add `chef_version`.
- README: Add github and license badges.

### Changed
- Update some cookbook versions:
  - `chef-vault` from `1` to `2`.
  - `dovecot` from `2` to `3`.
  - `postfixadmin` from `2` to `3` (fixes [issue #7](https://github.com/zuazo/postfix-dovecot-cookbook/issues/7), thanks [Arunderwood](https://github.com/arunderwood) for reporting).
  - `ssl_certificate` from `1` to `2`.
  - `onddo-spamassassin` from `1` to `2`.
  - `yum` from `3` to `5`.
- Dovecot: enable SSL explicitly.
- Update RuboCop to version `0.40` and fix new offenses.
- CHANGELOG: Follow "Keep a CHANGELOG".

### Removed
- Drop Chef `< 12.5` support.
- Drop Ruby `< 2.2` support.
- Metadata: Remove grouping ([RFC-85](https://github.com/chef/chef-rfc/blob/8d47f1d0afa5a2313ed2010e0eda318edc28ba47/rfc085-remove-unused-metadata.md)).
- Remove `postfix_full` recipe.
- README: Remove documentation about locale (old).

### Fixed
- Fix CentOS 7 and Scientific support with PostgreSQL.
- Fix PostgreSQL support on CentOS and Fedora.
- Fix Chef `13` deprecation warnings.

## [2.0.1] - 2015-09-03
### Fixed
- Fix typo in `-f` argument to `/usr/lib/dovecot/deliver` ([issue #5](https://github.com/zuazo/postfix-dovecot-cookbook/pull/5), thanks [Uwe Stuehler](https://github.com/ustuehler)).

## [2.0.0] - 2015-08-22
### Changed
- Update the `postfixadmin` cookbook to version `2` ([See the `postfixadmin` cookbook CHANGELOG for the update process](https://github.com/zuazo/postfixadmin-cookbook/blob/master/CHANGELOG.md#upgrading-from-a-1xy-cookbook-release)).
- Update `onddo-spamassassin` cookbook to version `1` ([See the `postfixadmin` cookbook CHANGELOG for the update process](https://github.com/onddo/spamassassin-cookbook/blob/master/CHANGELOG.md#v100-2015-04-29)).
- Update chef links to use *chef.io* domain.
- Update contact information and links after migration.
- Update RuboCop to `0.33.0`.

### Added
- metadata: Add `source_url` and `issues_url`.
- README: Put the cookbook name in the title.

## [1.2.0] - 2015-01-09
### Added
- Add Dovecot SSL certificate generation.
- Integrate with `ssl_certificate` cookbook version `1.2`.

### Changed
- Gemfile: Update RuboCop to `0.28.0`.

### Fixed
- metadata: Fix attributes default types.
- README: Fix some typos and update Supermarket links.

## [1.1.0] - 2014-11-09
### Added
- Add RBL support.
- Allow postfix configuration (tables and master.cf) to be modfied easily.
- Create Postfix tables directory, required by SES.

### Changed
- Improve SES support:
 - Read the SES credentials from chef vault bag.
 - Add `node['postfix-dovecot']['ses']['region']` attribute.
 - Update SES servers.
- metadata: update to use `dovecot` cookbook version `2`.
- Simplify `smtp_tls_CAfile` attribute case.
- Homogenize license headers.
- README:
  - Use single quotes in examples.
  - Use markdown tables.
  - Fix *Usage Examples* title.

### Fixed
- `::dovecot` recipe: Fix password reading with encrypt attributes enabled.
- Fix new RuboCop offenses.

## [1.0.0] - 2014-10-07
### Added
- Integrate with `ssl_certificate` cookbook.
- Add PostgreSQL support.

### Changed
- Rename `::postfix_full` recipe to `::postfix`.
- Update to work with `postfixadmin` cookbook `1.0.0`.
- Improve Postfix chroot file creation, based on `postfix-full` master code.
- Set `common_name` for PostfixAdmin and Postfix SSL certs.
- metadata: use pessimistic version constraints.
- Use `#default_unless` instead of `#set_unless`.
- `::postfixadmin` recipe: remove `#set_unless` usage.
- README:
  - Separate README file in multiple files.
  - Add some badges.
  - Some small documentation fixes.

### Removed
- Drop Ruby `< 1.9.3` support.
- Drop Chef `< 11.14.2` support.

### Fixed
- Fix hostname attribute default value when FQDN is not set.
- Fix all *RuboCop* offenses.

## [0.3.0] - 2014-09-14
### Added
- Ensure */etc/mailname* file creation.
- Add Fedora and Amazon Linux support.

### Changed
- Depends on `postfixadmin` cookbook version `< 1.0.0`.
- README: Amazon SES Tests section: KITCHEN_LOCAL_YAML variable value fixed.

## [0.2.0] - 2013-08-09
### Added
- Add [Amazon SES](http://aws.amazon.com/ses/) support.
  - Add SES tests.

### Fixed
- Fix *resolv.conf* inside chroot in CentOS.

## 0.1.0 - 2013-06-16

- Initial release of `postfix-dovecot`.

[Unreleased]: https://github.com/zuazo/postfix-dovecot-cookbook/compare/3.0.0...HEAD
[3.0.0]: https://github.com/zuazo/postfix-dovecot-cookbook/compare/2.0.1...3.0.0
[2.0.1]: https://github.com/zuazo/postfix-dovecot-cookbook/compare/2.0.0...2.0.1
[2.0.0]: https://github.com/zuazo/postfix-dovecot-cookbook/compare/1.2.0...2.0.0
[1.2.0]: https://github.com/zuazo/postfix-dovecot-cookbook/compare/1.1.0...1.2.0
[1.1.0]: https://github.com/zuazo/postfix-dovecot-cookbook/compare/1.0.0...1.1.0
[1.0.0]: https://github.com/zuazo/postfix-dovecot-cookbook/compare/0.3.0...1.0.0
[0.3.0]: https://github.com/zuazo/postfix-dovecot-cookbook/compare/0.2.0...0.3.0
[0.2.0]: https://github.com/zuazo/postfix-dovecot-cookbook/compare/0.1.0...0.2.0


metadata.rb
# encoding: UTF-8
#
# Cookbook Name:: postfix-dovecot
# Author:: Xabier de Zuazo (<xabier@zuazo.org>)
# Copyright:: Copyright (c) 2013-2015 Onddo Labs, SL.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

name 'postfix-dovecot'
maintainer 'Xabier de Zuazo'
maintainer_email 'xabier@zuazo.org'
license 'Apache 2.0'
description 'Installs and configures a mail server using Postfix, Dovecot, '\
            'PostfixAdmin and SpamAssassin, including Amazon SES support.'
long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version '3.1.0' # WiP

if respond_to?(:source_url)
  source_url "https://github.com/zuazo/#{name}-cookbook"
end
if respond_to?(:issues_url)
  issues_url "https://github.com/zuazo/#{name}-cookbook/issues"
end

chef_version '>= 12.5' if respond_to?(:chef_version)

supports 'amazon'
supports 'centos', '>= 6.0'
supports 'debian', '>= 7.0'
supports 'fedora', '>= 17.0'
supports 'ubuntu', '>= 12.04'

recipe 'postfix-dovecot::default', 'Installs and configures everything.'
recipe 'postfix-dovecot::vmail', 'Creates vmail user.'
recipe 'postfix-dovecot::spam', 'Installs and configures SpamAssassin.'
recipe 'postfix-dovecot::postfix', 'Installs and configures Postfix.'
recipe 'postfix-dovecot::postfix_mysql',
       'Installs Postfix package with MySQL support.'
recipe 'postfix-dovecot::postfix_postgresql',
       'Installs Postfix package with PostgreSQL support.'
recipe 'postfix-dovecot::postfixadmin', 'Installs and configures PostfixAdmin.'
recipe 'postfix-dovecot::dovecot', 'Installs and configures Dovecot 2.'

depends 'chef-vault', '~> 2.0'
depends 'build-essential', '~> 8.0'
depends 'dovecot', '~> 3.0'
depends 'onddo-spamassassin', '~> 2.0'
depends 'postfixadmin', '~> 3.0'
depends 'postfix-full', '~> 0.1'
depends 'ssl_certificate', '~> 2.0'
depends 'yum', '~> 5.0'

attribute 'postfix-dovecot/postmaster_address',
          display_name: 'postmaster address',
          description: 'Postmaster mail address.',
          type: 'string',
          required: 'recommended',
          default: 'postmaster@foo.bar'

attribute 'postfix-dovecot/rbls',
          display_name: 'postfix rbls',
          description: 'Mail RBLs array.',
          type: 'array',
          required: 'recommended',
          default: []

attribute 'postfix-dovecot/hostname',
          display_name: 'hostname',
          description: 'Hostname.',
          type: 'string',
          required: 'recommended',
          calculated: true

attribute 'postfix-dovecot/database/type',
          display_name: 'postfix database type',
          description: 'Postfix database type.',
          choice: %w(mysql postgresql),
          type: 'string',
          required: 'optional',
          default: 'mysql'

attribute 'postfix-dovecot/sieve/enabled',
          display_name: 'sieve enabled',
          description: 'Whether to enable sieve.',
          type: 'string',
          choice: %w(true false),
          required: 'recommended',
          default: 'true'

attribute 'postfix-dovecot/sieve/global_path',
          display_name: 'sieve global_path',
          description: 'Sieve global path.',
          type: 'string',
          required: 'optional',
          calculated: true

attribute 'postfix-dovecot/spamc/enabled',
          display_name: 'spamc enabled',
          description: 'Whether to enable SpamAssassin.',
          type: 'string',
          choice: %w(true false),
          required: 'optional',
          default: 'true'

attribute 'postfix-dovecot/spamc/recipe',
          display_name: 'spamc recipe',
          description: 'Spamc recipe name to use.',
          type: 'string',
          required: 'optional',
          default: 'onddo-spamassassin'

attribute 'postfix-dovecot/vmail/user',
          display_name: 'vmail user',
          description: 'Virtual mail system user name.',
          type: 'string',
          required: 'optional',
          default: 'vmail'

attribute 'postfix-dovecot/vmail/group',
          display_name: 'vmail group',
          description: 'Virtual mail system group name.',
          type: 'string',
          required: 'optional',
          calculated: true

attribute 'postfix-dovecot/vmail/uid',
          display_name: 'vmail uid',
          description: 'Virtual mail system user id.',
          type: 'string',
          required: 'optional',
          default: '5000'

attribute 'postfix-dovecot/vmail/gid',
          display_name: 'vmail gid',
          description: 'Virtual mail system group id.',
          type: 'string',
          required: 'optional',
          calculated: true

attribute 'postfix-dovecot/vmail/home',
          display_name: 'vmail home',
          description: 'Virtual mail user home path.',
          type: 'string',
          required: 'optional',
          default: '/var/vmail'

attribute 'postfix-dovecot/ses/enabled',
          display_name: 'ses enabled',
          description: 'Whether to enable Amazon SES.',
          type: 'string',
          choice: %w(true false),
          required: 'recommended',
          default: 'false'

attribute 'postfix-dovecot/ses/source',
          display_name: 'ses credentials source',
          description: 'Where to read the credentials from.',
          type: 'string',
          choice: %w(attributes chef-vault),
          required: 'recommended',
          default: 'attributes'

attribute 'postfix-dovecot/ses/vault',
          display_name: 'ses credentials vault',
          description: 'Chef Vault bag to read SES credentials from.',
          type: 'string',
          required: 'recommended',
          default: 'amazon'

attribute 'postfix-dovecot/ses/item',
          display_name: 'ses credentials vault item',
          description: 'Chef Vault item.',
          type: 'string',
          required: 'recommended',
          default: 'ses'

attribute 'postfix-dovecot/ses/region',
          display_name: 'ses region',
          description: 'Amazon AWS region, used to calculate the servers.',
          type: 'string',
          required: 'optional',
          default: 'us-east-1'

attribute 'postfix-dovecot/ses/servers',
          display_name: 'ses servers',
          description: 'Amazon SES SMTP servers.',
          type: 'array',
          required: 'optional',
          default: %w(
            email-smtp.us-east-1.amazonaws.com:25
            ses-smtp-prod-335357831.us-east-1.elb.amazonaws.com:25
          )

attribute 'postfix-dovecot/ses/username',
          display_name: 'ses username',
          description: 'Amazon SES SMTP username.',
          type: 'string',
          required: 'recommended',
          default: 'USERNAME'

attribute 'postfix-dovecot/ses/password',
          display_name: 'ses password',
          description: 'Amazon SES SMTP password.',
          type: 'string',
          required: 'recommended',
          default: 'PASSWORD'

attribute 'postfix-dovecot/yum',
          display_name: 'yum repositories',
          description:
            'A list of yum repositories to add to include the source SRPMs.',
          type: 'hash',
          required: 'optional',
          calculated: true

attribute 'postfix-dovecot/srpm/packages',
          display_name: 'srpm packages',
          description: 'Packages required for compiling Postfix from sources.',
          type: 'array',
          required: 'optional',
          calculated: true

attribute 'postfix-dovecot/srpm/rpm_regexp',
          display_name: 'srpm rpm regexp',
          description:
            'An array with two values, a pattern and a replacement. This'\
            'Regexp is used to get the final Postfix RPM name from the SRPM '\
            'name.',
          type: 'array',
          required: 'optional',
          calculated: true

attribute 'postfix-dovecot/srpm/rpm_regexp',
          display_name: 'srpm rpm regexp',
          description:
            'A string with the arguments to pass to rpmbuild application. '\
            'Normally contains the required option to enable PostgreSQL in '\
            'the Postfix SRPM.',
          type: 'hash',
          required: 'optional',
          calculated: true
